---
title: "Outline Of Analysis"
---

```{r set-knitr-opts, echo=FALSE, child = '../workflow/modules/setup_chunk.Rmd'}
```


```{r packages}
library(tidyverse)
library(DiagrammeR)
library(pander)
library(yaml)
library(scales)
```

```{r remotes, results='hide'}
## Deal with github packages
if (!"spBioUtils" %in% rownames(installed.packages()))
  BiocManager::install("steveped/spBioUtils")
stopifnot(library(spBioUtils, logical.return = TRUE))
```

```{r config}
config <- here::here("config/config.yml") %>%
  read_yaml()
samples <- config$samples %>%
  here::here() %>%
  read_tsv()
```


## Description

This is a standardised workflow for data preparation and a naive analysis of RNA-Seq data.
Given that all steps are automated, **results must be checked carefully**.
Multiple plots are provided for quick visual identification of any potential problems.

The default analysis will perform one or more pairwise comparisons, as specified in `config/config.yml`.
The workflow is presented as a DAG below, along with fine details

## Steps

### QC

This is performed on raw and trimmed files using [`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), with summaries written using [`ngsReports`](https://bioconductor.org/packages/release/bioc/html/ngsReports.html)
QC steps on count data is also provided.

All of these steps are executed regardless of any comparisons being performed.

### Read Trimming

This is performed using [`AdapterRemoval`](https://adapterremoval.readthedocs.io/en/stable/) with parameters able to be provided via `config.yml`

### Differential Expression

The approach here is to summarise to gene-level counts prior to analysis.

- Data is normalised using [Condition-Quantile normalisation](https://doi.org/10.1093/biostatistics/kxr054) accounting for bias from GC content or gene length.
Only detected (i.e. non-zero) transcripts are used in the estimation of these values.
- The [Quasi-Likelihood approach](https://doi.org/10.1515/1544-6115.1826) is used for dispersion estimates
- Hypothesis testing uses a range-based $H_0$ via [`glmTreat()`](https://doi.org/10.1093/bioinformatics/btp053)


## DAG

```{r plot-workflow, fig.height = 10, fig.width = 10, fig.cap = "*Summary of processing workflow. The primary data pipeline is shown in red, preparatory steps are shown in blue whilst collation of final output is in green.*"}
here::here("workflow", "rules", "rulegraph.dot") %>%
  readLines() %>%
  rm_dot_node(node = "\"all\"") %>%
  add_input_node(node = "Raw Data", col = "red", ignore = "(download|define|create|make)") %>%
  change_node_colour("(remove|salmon|counts|analysis)", "red") %>%
  change_node_colour("(download|make|index)", "blue") %>%
  change_node_colour("(find|fastqc|site|template)", "forestgreen") %>%
  str_replace_all("_", "\n") %>%
  str_replace_all("snakemake\ndag", "snakemake_dag") %>%
  str_replace_all("fontsize=[0-9]+", "fontsize=16") %>%
  str_replace_all("(.+graph.+)(\\];)", "\\1, rankdir = LR\\2") %>%
  grViz()
```

## Parameter Description

Parameters set for the complete workflow are:

```{r printYAML, echo=FALSE, results='asis'}
pander(config)
```

## Samples

The samples provided in the config file are:

```{r print-samples, echo=FALSE}
pander(samples)
```


<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
```{r session-info, echo=FALSE}
pander::pander(sessionInfo())
```
</div>
